### PR TITLE
chore(a11y): standardize keyboard focus

### LINKED ISSUE #48121: (a11y) Keyboard focus indicator
### Describe the Issue

We are relying on the browser's default keyboard focus indicator for quite a few focusable elements and it just isn't very accessible in certain browsers (looking at you Chrome). We really need to settle on a focus indicator color and then make sure every focusable element uses it for the focus outline. I've been using our CSS custom variable `--blue-mid` for things I've worked on because it has accessible contrast with both our light and dark themes (you can see this in the main nav menu or the tabs when viewing a step in narrow view). 

### Affected Page

https://www.freecodecamp.org/learn/2022/responsive-web-design/learn-typography-by-building-a-nutrition-label/step-66

### Steps to Reproduce

Using Chrome (or any chrome-based browser), open the page in wide (desktop) view and start tabbing through the controls. When you get to the five buttons above the editor the focus outline will barely be noticeable on some of them and pretty much non-existent on others. Also notice that we have several different focus outline colors in the header/breadcrumbs. While not technically a WCAG violation I think having a standard color for all of them would be better (not just aesthetically but also for accessibility).

### Expected behavior

Perhaps create a custom CSS variable for the focus outline color and then start updating all of our focusable elements to use it. The width of the outline could vary depending on how much space we have and where the element is located, but I would recommend at least a width of 2px. 

Even if we won't be able to settle on an official outline color for a while, I still think we should pick a temporary one for now and put it in place so that when we do finally settle on a color it will be relatively easy to make the switch. Also, CSS allows us to only show the focus outline to people using the keyboard to navigate the page, so I'm guessing most of our users won't even see it. Thus, picking a temporary color for now isn't going to affect the aesthetics for most users but will make an immediate impact on accessibility for those who rely on it. I guess I'm asking that we make the accessibility improvements now instead of waiting until we have an official color scheme in place.

### Screenshots

_No response_

### System

- Device: Desktop
- OS: Both Linux Fedora 36 and Windows 11
- Browser: Chrome
- Version: 106

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by a2937**
Checklist:

- [X] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [X] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [X] My pull request targets the `main` branch of freeCodeCamp.
- [X] I have tested these changes either locally on my machine, or GitPod.

Closes #48121

Same color for both the dark and light color palettes since it is a medium-color.  It was actually pretty easier adding the variable to the css variable list and the focus selector to globals.css.

### CODE REVIEW

File: client/src/components/layouts/global.css
Comment (Sboonny) [inline client/src/components/layouts/global.css:192]: You don't need to create `--focus-outline-color` if the color is the same in both themes.



---

This selector isn't applied, looking at the devtools it doesn't reach the interactive elements, because of bootstrap higher specificity class.

The outline doesn't have enough offset like mentioned in the issue.

if you fix the specificity issue from bootstrap, can you remove the elements that have the same outline styles to clear some of the bloating css.

Comment (a2937) [reply client/src/components/layouts/global.css:192]: I don't remember seeing an offset mentioned in the issue? Can you point it out? I remember the width of the outline being specified though.

Comment (Sboonny) [reply client/src/components/layouts/global.css:192]: True, I have read the issue again and it doesn't include offset, you can ignore offset part.

Comment (a2937) [reply client/src/components/layouts/global.css:192]: Additionally in hindsight , should I be using using border-color over outline in the css?

Comment (Sboonny) [reply client/src/components/layouts/global.css:192]: border-color has width and height that are respected, which can break the layout and shift items in the page. outline on the other had doesn't have respected width and height.

Comment (a2937) [reply client/src/components/layouts/global.css:192]: Pardon me Sboony, but are you sure my focus selector isn't being applied? 

It appears that focus selectors don't actually show up in the inspector.

Example

There's also a JS fiddle I want to show you locally as a reference to what I'm talking about.

Comment (Sboonny) [reply client/src/components/layouts/global.css:192]: can you be more specific of what is the question?

Comment (a2937) [reply client/src/components/layouts/global.css:192]: I was trying to point that `:focus` selectors don't show up in the inspector and was trying to point out the possibility that maybe they're being applied but we can't easily find it out other than by hitting tab and seeing them the outline show up by hitting tab. 

My question was how are we supposed to be one hundred percent that my selector wasn't being applied. Like is there a way to find out using dev tools or not?

**Comment by bbsmooth** [comment]
I am assuming we only want the keyboard focus indicator to show if the user is actually using a keyboard? That's the way I've been implementing the focus indicator as I've been fixing accessibility issues.

If so, then this needs to target the `:focus-visible` pseudo-class instead of `:focus`. 

`:focus-visible` has been around long enough in the major browsers/platforms that we probably don't need to worry about a fallback for the minority of browsers that don't support it. But if we do want to be extra careful then we can use `@supports` to target those browsers.

[MDN: focus-visible](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible#providing_a_focus_fallback)

File: client/src/components/layouts/global.css
Comment (Sboonny) [reply client/src/components/layouts/global.css:192]: In the css section of the devtools there are `:hov` button next to the search click it and click `:focus` or `:focus-visible` if you have changed it already.

Whatever interactive element you select will have focus indicator, your change isn't applied because of its specificity.


Yes you are technically applying the outline, but what is the point if the applied change isn't used, because `class:focus` elements which have the issue have a higher specificity, so your change is ignored when it's needed.

Comment (a2937) [reply client/src/components/layouts/global.css]: Thank you for teaching me something new.

**Comment by a2937** [comment]
My selector being applied

File: client/src/components/layouts/global.css
Comment (a2937) [reply client/src/components/layouts/global.css]: Okay where is it not being applied where it should? That will help narrow down my search.

File: client/src/components/layouts/variables.css
Comment (bbsmooth) [inline client/src/components/layouts/variables.css:45]:
```suggestion
  --focus-outline-color: var(--blue-mid);
```

The outline color for both light/dark needs to stay with blue-mid. This color of blue meets color contrast requirements for both themes at the same time. blue-light has issues in the dark theme and blue-dark has issues in the light theme.

File: client/src/components/layouts/variables.css
Comment (bbsmooth) [inline client/src/components/layouts/variables.css:66]:
```suggestion
  --focus-outline-color: var(--blue-mid);
```

File: client/src/components/layouts/variables.css
Comment (a2937) [reply client/src/components/layouts/variables.css:45]: Okay but because they are the same, could I simply define it at the root?

**Comment by bbsmooth** [comment]
The way I saw this working would be that we define a default outline color (--blue-mid), outline width (3px), outline style (solid), and outline offset (0) (which would be the same for both themes), and then we can make the necessary adjustments for specific situations. I think it would be safe to set these all the root.

As for the adjustments, I don't think the color will ever need to change (although I could be wrong about this) except for one specific case, the steps grid in dark mode. Because we are using a blue background to mark completed steps the outline needs to be white. I think most of the adjustments will need to be made to the offset, and possible width. These adjustments could be different for each theme.

There are a lot of focusable elements, so this PR is potentially going to take a while. @a2937, do you mind if I help out and push changes directly to this PR?

**Comment by a2937** [comment]
I don't mind the additional assist.

File: client/src/templates/Challenges/classic/classic.css
Comment (bbsmooth) [inline client/src/templates/Challenges/classic/classic.css:117]: I think we can delete `outline` completely since it is the same as the default.

Comment (a2937) [reply client/src/templates/Challenges/classic/classic.css:117]: It's a catch for the focus-visible property not existing. We aren't going to have an outline if the `:focus-visible` isn't supported on that particular browser. Is this something everyone is okay with? 

That being said, I could very easily delete if we change the selector I made to a regular focus one.

Comment (bbsmooth) [reply client/src/templates/Challenges/classic/classic.css:117]: Yes, it's a fallback, but in this case the fallback is specific to `#mobile-layout-pane-instructions`. I have no problem having a global fallback just in case, but if we are going to have that then it should be in global.css right below where we create the default `:focus-visible` outline. Then the only thing we need in this fallback is the offset.

File: client/src/components/layouts/global.css
Comment (Sboonny) [reply client/src/components/layouts/global.css]: In https://www.freecodecamp.org/learn/2022/responsive-web-design/learn-typography-by-building-a-nutrition-label/step-66, you can find a couple of areas that still have different outline colors cased by bootstrap.

> Also notice that we have several different focus outline colors in the header/breadcrumbs. While not technically a WCAG violation I think having a standard color for all of them would be better (not just aesthetically but also for accessibility).

image

> pretty much non-existent on others. 

These are in the setting page, any toggle button doesn't have any focus indicator.

**Comment by Sboonny** [comment]
> I'm guessing that client/src/components/layouts/global.css is where we would want to define the default? @Sboonny, correct me if I'm wrong?

I am happy with it in the global.css file 👍

File: client/src/components/layouts/global.css
Comment (Sboonny) [reply client/src/components/layouts/global.css]: To save time, because we want to apply this everywhere with the same color and stuff and maybe delete the classes in bootstrap, it will break the outline in specific location, but it's better to add specific fixes, than try to overwrite everything, and overwrite it again whenever something new is added.

File: client/src/templates/Challenges/classic/classic.css
Comment (a2937) [reply client/src/templates/Challenges/classic/classic.css]: Okay I believe I just resolved this

**Review by Sboonny — APPROVED** [review]
@a2937, thank you for hard work, your changes has highlighted issue with the logo, I am planning to fix, after the fix this is good to go

**Comment by ahmadabdolsaheb** [comment]
Could you rewrite the change to bootstrap.min.css instead of altering the file?

**Comment by Sboonny** [comment]
What do you mean by overwrite? 

---

I was the one to suggest deleting bootstrap.css https://github.com/freeCodeCamp/freeCodeCamp/pull/49579#discussion_r1127404349, because:

- trying to increase the specificity over bootstrap value will create weird ids, that will force terrible developer experience whenever someone interacts with the outline, as it will get in the way whether bootstrap or the new add class.
- using `!important` will break the outline in important section like the grid for the challenges in the block map
- trying to work around it provide more trouble than it's worth, and it is getting on the way.

**Comment by ahmadabdolsaheb** [comment]
Thanks for providing more context. The minified files are for consumption only as the styles context such as bootstrap main, normalization, customization, etc. is removed during magnification.

It works for me if you are on board with it. It would be great if you could share your review process for the minified file so I could give it a proper review before merge.

**Comment by Sboonny** [comment]
- Open `bootstrap.min.css` file, format it with CSS not prettier, 
- git add -A
- Copy raw text from `bootstrap.min.css` file, in the main repo, here is it https://raw.githubusercontent.com/freeCodeCamp/freeCodeCamp/main/client/static/css/bootstrap.min.css
- paste it inside the file in this branch, and format it with CSS, but don't add it.

When done correctly, you should see git diff and compare what is deleted from the main bootstrap

image

**Comment by bbsmooth** [comment]
@Sboonny @ahmadabdolsaheb A quick question. Is it the intent to make sure all focus indicators are functioning the way we want them to in this PR? Or is the intent just to get the default in place and then in other PRs we can fix any that still aren't working correctly?

**Comment by Sboonny** [comment]
I have missed the comment about adding a commit, sorry for early approval.

> is the intent just to get the default in place

I have reviewed this PR based, on setting default in place, I have known about logo issue, before the PR was opened, and I wouldn't have wanted this PR to include these fixes.

But this was something that can be solve without the need for the PR to merged, so I was comfortable with it, if you see an issue that need to be fixed in this PR, feel free to push the commit.

**Review by ahmaxed — APPROVED** [review]
LGMT upon resolving the conflicts.

**Comment by bbsmooth** [comment]
To expand on my earlier comment. For example, in the steps, the focus outline on both the reset and ask for help buttons are not showing properly. Do we want to fix that in this PR? I haven't gone through every focusable element to ensure that the outline is showing properly. Is this something we want to verify before we merge this PR? Or should we just get these defaults merged now and then I will go through and fix any minor issues I find in separate PRs? 

Also, I asked in the issue if we wanted the outline to wrap around the borders of buttons so that the border still shows or if we wanted the outline to cover the border. @ahmadabdolsaheb leaned toward covering the border, and so do I. So do we want to take care of that in this PR as well?

**Comment by Sboonny** [comment]
I have investagated it more, and the lower-jaw buttons can be fixed, by removing the extra `overflow` property, if removing the property from `.lower-jaw-icon-bar` is all you plan to do with lowerjaw button, and make the outline cover the border feel free to push a commit, I will re-review it and merge it👍,

If you are planning to do more, like cover the border of the lowerjaw button too, maybe it's better to do it in different PR, and add the text for the buttons 🤔.

Either way this shouldn't be merged with that lower-jaw's buttons looks

**Comment by Sboonny** [comment]
btw, the logo is broken 😞, my fix didn't help much. 

I will create another PR turning the layout from flex to grid, and be done with it once and for all.

**Comment by bbsmooth** [comment]
I pushed two fixes, one for the icon buttons in the lower jaw and one for the bread crumb links. I don't think I'm seeing any other blockers. There is some minor tweaking that probably needs to be done but that can be handled in separate PRs.

I'm going to hold of on officially reviewing this just a little longer as I really want to go through everything one last time but I can't do that until later this evening.

**Review by bbsmooth — APPROVED** [review]
I am not finding any blockers that should prevent merging this PR. There are some minor issues I will address in separate PRs.

For what it's worth, you have my LGTM.

**Pull request merged**
